Add filter to remove elements by CSS rule from HTML before change detection is run

timlod opened the pull request:
This PR adds a filter that will exclude header, footer and nav tags from HTML.

This is useful in situations where it is expected that interesting changes only happen outside of header, footer or navigation (in most cases that means in the main body of the document).

I get some failing tests, but tbh I have no clue why - I'm adding something very similar to PR #444, which passes, and the failing tests don't seem to be related directly to my changes.

The tests I added for this work (which pass) I adapted from `test_css_selector.py` and am not 100% sure about `test_body_filter_full` - @dgtlmoon perhaps you can have a closer look to make sure my intention with the test is executed correctly?

dgtlmoon commented:
What would be even better is a configurable list

With a link to auto-add the common ones
I can imagine people will want to try use this in the future

And we could use xpath or css rules to ignore parts in the same textarea, more flexible

this feels more future-proof

![image](https://user-images.githubusercontent.com/275001/156554268-42bdc91c-7291-4402-8cc5-905fdf9ef4fe.png)


is it possible? :)

dgtlmoon commented:
Maybe *Ignore Elements* is better

![image](https://user-images.githubusercontent.com/275001/156554414-3e0d4dc4-90a0-4182-a715-1503de1684e6.png)

dgtlmoon commented:
Yeah in this way, it solves two PR's in one https://github.com/dgtlmoon/changedetection.io/pull/444

timlod commented:
@dgtlmoon I implemented your suggestion, bar the fact that currently XPath subtractive filters are not supported, and that I add suggestions in the placeholder of the form rather than adding links to add header/footer/nav. These could be added though if you think that's a better solution!

dgtlmoon commented on changedetectionio/html_tools.py:
```diff
@@ -1,7 +1,10 @@
 import json
+import re
+from typing import List
```
new to me! interesting

timlod replied:
Python type hints (check out https://docs.python.org/3/library/typing.html) To be honest, it's a bit odd that I added this single one - using them across the whole project would be cool though, could make development easier as IDEs can leverage them and inform you if expected input is violated.

dgtlmoon commented on changedetectionio/templates/edit.html:
```diff
@@ -116,6 +116,16 @@
+nav
+.stockticker") }}
+                      <br/>
+                      Ignore HTML tags or matches to CSS selectors.
```
I think just only supporting CSS here (no xpath etc) is totally fine for now

dgtlmoon replied:
@timlod please wrap this with

```
<span class="pure-form-message-inline">
```

like the other "text" in that form

timlod commented:
I'm not sure why the test is failing - it passed for me locally. Any clues?

dgtlmoon commented:
> I'm not sure why the test is failing - it passed for me locally. Any clues?

Pass local here too.. are you able to trigger a re-run of the tests? I just triggered a re-run in anycase

dgtlmoon commented on changedetectionio/forms.py:
```diff
@@ -1,13 +1,30 @@
 
-from wtforms import widgets, SubmitField
-from wtforms.validators import ValidationError
+from wtforms import (
```
thanks for cleaning this stuff up

timlod replied:
This is the work of `isort` (https://github.com/PyCQA/isort) running automatically on every save!
It may make sense to run it (and perhaps also some code formatter) across the whole project and mandate it in the contributing file, that would make contributing and keeping the codebase uniform easier.

dgtlmoon replied:
@timlod I made a wiki here https://github.com/dgtlmoon/changedetection.io/wiki/Developer-Notes could you update this with some of the very cool tools you're using? I'm finding it hard to keep up :)

dgtlmoon commented:
I'm wondering if this should be called "Strip elements" rather than "Ignore elements" ? because it's stripping the element from the HTML before checksum/trigger/etc is running right?

dgtlmoon commented:
`subtractive_filters` it's not a filter per-se, its more like removing it from the DOM

`subtractive_filters` => `element_removal` ?

dgtlmoon commented:
Also `filters` show up in the "preview" screenshot as a highlighted filter

![image](https://user-images.githubusercontent.com/275001/156769246-f9e9bb12-2ff6-4c2a-8f61-25eefa212022.png)

but this is more like removing it entirely? or not?

timlod commented:
You're right, perhaps making the naming here clearer would be good.
Removal is fine, I will commit that later.

About the preview section, I haven't really investigated how that works so I'm not sure what you mean.
I thought that the removed elements should not show up in preview/diff at all, as they should be removed before the text is saved or used for the trigger. Could that be a leftover from a site saved before applying the removal?

dgtlmoon commented:
@timlod please read https://github.com/dgtlmoon/changedetection.io/issues/425

basically the guys were saying "it doesnt work" but there was no way to visualise the filter, filter does not _remove_ but just skips that text as part of the checksum

dgtlmoon commented:
so this filter is filter _after_ HTML conversion

your _remover_ is _before_ HTML conversion

dgtlmoon commented:
almost need some cute SVG diagram to show the process to people :)

timlod commented:
Ah, sorry, I misunderstood your previous comment (with the screenshot).
Yes, that's exactly how I thought it worked. I thought you were pointing our that my _remover_ was showing up as a highlighted filter. Indeed it's working just as intended.

dgtlmoon commented:
Ok so only remaining thing for me here is just to tweak the terminology, then I'll merge it in I think

`element_removal` ? unsure exactly

dgtlmoon commented:
I would like to get this into the next release `0.39.10` I think

timlod commented:
I'll try to do these last bits this evening.

timlod commented:
OK, I added the span to the HTML and renamed the method to `element_removal`. Further I refactored the subtractive filters to be called subtractive selectors, which more accurately represents what they are. How do you feel about also renaming the `css_filter` (the one given in `forms.py` which would be used for the `html_tools` function `css_filter`) to `css_selector`?

dgtlmoon commented:
`css_selector` would be a nicer name but i'm afraid to change the users JSON for this structure atm

we don't have a good way to handle "schema upgrade versions" yet..

although, yeah, feel free to open a new issue I think, would be a good one

dgtlmoon commented on changedetectionio/tests/test_element_removal.py:
```diff
@@ -0,0 +1,164 @@
+
+    # It should have 'unviewed' still, as we removed changed elements
+    res = client.get(url_for("index"))
+    assert b"unviewed" in res.data
```
In  your test, you are for example ignoring `filter`

Would be really nice to have a test that has just `<footer> something new</footer>` and then 

```
res = client.get(url_for("index"))
assert b"unviewed" not in res.data
```

because in this test, it doesn't really prove that it wasnt the element-removal that caused the `unviewed` to be shown.. if you understand?

dgtlmoon replied:
extra points if the test can also set the `css_filter` of `footer` too :) :)

timlod replied:
Not sure what you mean here exactly - I've changed the tests now such that there are changes both in header, nav, footer as well as the selected css class - none of which trigger an unviewed change.

dgtlmoon commented:
by the way - whats the purpose of using this 'remove' versus the existing css filter? as they essentially do the same thing? (stop a change alert from happening)

timlod commented:
> by the way - whats the purpose of using this 'remove' versus the existing css filter? as they essentially do the same thing? (stop a change alert from happening)

The existing CSS filter selects and matches only a specific element. The goal here is the opposite - remove specific elements that throw false positives.

dgtlmoon commented:
> The existing CSS filter selects and matches only a specific element. The goal here is the opposite - remove specific elements that throw false positives.

yes thats right :)

sorry, brain fail

dgtlmoon commented:
https://github.com/dgtlmoon/changedetection.io/pull/445#discussion_r820295680  just needs this one

timlod commented:
Sorry, thought test is passing, turns out another webservice I had running on port 5005 made everything pass :) Will fix this now.

timlod commented:
Now it's that same test that was failing previously, same as at #431. @dgtlmoon could you shed some light on what may happen here?

dgtlmoon commented:
@timlod could you try commit this patch to your branch
[fix-445-patch.txt](https://github.com/dgtlmoon/changedetection.io/files/8216801/fix-445-patch.txt)

timlod commented:
That did not work, lots of errors like these:
`changedetectionio:update_worker.py:63 Exception reached processing watch UUID: 90515af6-3bc2-4fb8-8ab9-f162407d1b8e - HTTPConnectionPool(host='localhost', port=5005): Max retries exceeded with url: /test-body (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f21f44709a0>: Failed to establish a new connection: [Errno 111] Connection refused'
))`

dgtlmoon commented:
@timlod think i found it! https://github.com/dgtlmoon/changedetection.io/pull/453

pull that into your branch - dont know why your PR seemed to tickle this bug a bit more tho

dgtlmoon commented on changedetectionio/fetch_site_status.py:
```diff
@@ -1,11 +1,13 @@
+
 import urllib3
+from inscriptis import get_text
+
```
some minor white-lines here could be removed

timlod replied:
There's logic to this - there will be newlines between standard library, external libraries, and internal libraries. I did notice that there was a duplicate import of `html_tools` though, which I'll remove.

dgtlmoon replied:
Super cool! I'm learning a lot from you :dancers:

dgtlmoon commented on changedetectionio/templates/settings.html:
```diff
@@ -83,7 +83,18 @@
+.stockticker") }}
+                      <span class="pure-form-message-inline">
+                      <br/>
+                      Ignore HTML tags or matches to CSS selectors.
```
This should be "Remove HTML element by CSS selector before text conversion"

dgtlmoon replied:
Also better is to wrap this like 
```
<UL><LI>    Ignore HTML tags or matches to CSS selectors.</li>
<LI>Add one tag or CSS selector per line to ignore multiple parts of the HTML.</LI>
...
```

same like the other existing form entries

dgtlmoon commented on changedetectionio/templates/settings.html:
```diff
@@ -83,7 +83,18 @@
+                      <br/>
+                      Ignore HTML tags or matches to CSS selectors.
+                      <br/>
+                      Add one tag or CSS selector per line to ignore multiple parts of the HTML.
```
"Add multiple elements or CSS selectors per line to ignore multiple parts of the HTML." is better

dgtlmoon commented on changedetectionio/forms.py:
```diff
@@ -283,6 +319,7 @@ class watchForm(commonSettingsForm):
     minutes_between_check = html5.IntegerField('Maximum time in minutes until recheck',
                                                [validators.Optional(), validators.NumberRange(min=1)])
     css_filter = StringField('CSS/JSON/XPATH Filter', [ValidateCSSJSONXPATHInput()])
+    subtractive_selectors = StringListField('Ignore elements', [ValidateCSSJSONXPATHInput(allow_xpath=False, allow_json=False)])
```
"Remove elements"

dgtlmoon commented:
@timlod looks like one or two small merge conflicts for you, then I will cleanup the commit message in the squash/merge

dgtlmoon commented:
wont let me merge :(

dgtlmoon commented:
this is cool, I'm already using it!

dgtlmoon commented:
really good for noisy menus
